Source/Core/Core/HW/EXI/EXI_DeviceBaseboard.cpp

```diff
@@ -25,33 +116,189 @@ bool CEXIBaseboard::IsPresent() const
   return true;
 }
 
+bool CEXIBaseboard::IsInterruptSet()
+{
+  if (s_interrupt_set)
+  {
+    DEBUG_LOG_FMT(SP1, "AM-BB: IRQ");
+    if (++s_irq_timer > 12)
+      s_interrupt_set = false;
+    return true;
+  }
+
+  return false;
+}
+
+void CEXIBaseboard::DMAWrite(u32 addr, u32 size)
+{
+  const auto& system = Core::System::GetInstance();
+  const auto& memory = system.GetMemory();
+
+  NOTICE_LOG_FMT(SP1, "AM-BB: COMMAND: Backup DMA Write: {:08x} {:x}", addr, size);
+
+  m_backup.Seek(m_backup_offset, File::SeekOrigin::Begin);
+  m_backup.WriteBytes(memory.GetSpanForAddress(addr).data(), size);
+  m_backup.Flush();
+}
+
+void CEXIBaseboard::DMARead(u32 addr, u32 size)
+{
+  const auto& system = Core::System::GetInstance();
+  const auto& memory = system.GetMemory();
+
+  NOTICE_LOG_FMT(SP1, "AM-BB: COMMAND: Backup DMA Read: {:08x} {:x}", addr, size);
+
+  m_backup.Seek(m_backup_offset, File::SeekOrigin::Begin);
+  m_backup.ReadBytes(memory.GetSpanForAddress(addr).data(), size);
+}
+
 void CEXIBaseboard::TransferByte(u8& byte)
 {
-  if (m_position == 0)
+  DEBUG_LOG_FMT(SP1, "AM-BB: > {:02x}", byte);
+  if (m_position < 4)
   {
-    m_command = byte;
+    m_command[m_position] = byte;
+    byte = 0xFF;
   }
-  else
+
+  if ((m_position >= 2) && (m_command[0] == 0 && m_command[1] == 0))
+  {
+    // Read serial ID
+    byte = "\x06\x04\x10\x00"[(m_position - 2) & 3];
+  }
+  else if (m_position == 3)
   {
-    switch (m_command)
+    u32 checksum = (m_command[0] << 24) | (m_command[1] << 16) | (m_command[2] << 8);
+    u32 bit = 0x80000000UL;
+    u32 check = 0x8D800000UL;
+    while (bit >= 0x100)
     {
-    case 0x00:
+      if (checksum & bit)
+        checksum ^= check;
+      check >>= 1;
+      bit >>= 1;
+    }
+
+    if (m_command[3] != (checksum & 0xFF))
+      DEBUG_LOG_FMT(SP1, "AM-BB: cs: {:02x}, w: {:02x}", m_command[3], checksum & 0xFF);
+  }
+  else if (m_position == 4)
+  {
+    switch (m_command[0])
     {
-      static constexpr std::array<u8, 4> ID = {0x06, 0x04, 0x10, 0x00};
-      byte = ID[(m_position - 2) & 3];
+    case BackupOffsetSet:
+      m_backup_offset = (m_command[1] << 8) | m_command[2];
+      DEBUG_LOG_FMT(SP1, "AM-BB: COMMAND: BackupOffsetSet:{:04x}", m_backup_offset);
+      m_backup.Seek(m_backup_offset, File::SeekOrigin::Begin);
```

Comment:
This is an entirely different topic than the security review I'm focusing on, but: Right now, this seek offset isn't being savestated.